### DISCUSSION TITLE
Actions logs: relative order of messages in stdout and stderr is not preserved

### DISCUSSION THREAD

**Discussion by RalfJung** [community/community#116552] [Actions]
### Select Topic Area

Bug

### Body

We have a script like this running in Github Actions:
```
echo "::group::Display CPU and Memory information"
[...]
echo "::endgroup::"

if [ ! -z "$SCRIPT" ]; then
  echo "Executing ${SCRIPT}"
  sh -x -c "$SCRIPT" # `-x` will print each step of the script to stderr as `+ command`
else
[...]
```
And we are seeing output like this:
```
2024-03-31T16:59:56.6924564Z ##[group]Display CPU and Memory information
[...]
2024-03-31T16:59:56.7335902Z + /tmp/checktools.sh ../x.py
2024-03-31T16:59:56.7336189Z + X_PY=../x.py
2024-03-31T16:59:56.7336415Z + set +e
2024-03-31T16:59:56.7337412Z + python3 ../x.py test --stage 2 --no-fail-fast src/doc/book src/doc/nomicon src/doc/reference src/doc/rust-by-example src/doc/embedded-book src/doc/edition-guide
2024-03-31T16:59:56.7338518Z ##[endgroup]
2024-03-31T16:59:56.7339318Z Executing /tmp/checktools.sh ../x.py &&   NODE_PATH=`npm root -g` python3 ../x.py test tests/rustdoc-gui --stage 2     --test-args '--no-sandbox --jobs 1'
```
Notice how in the script, "endgroup" and "Executing ..." are emitted *before* `sh -x -c ...`, but in the log they appear *after* the first `+ command` lines! In other words, the log is just wrong about the order on which these lines happened. (The timestamps therefore are also wrong, but that's less important.)

This is pretty bad as it makes debugging actions very hard. I can't even trust that if A happens before B then I will see the log from A before the log from B! It gets even worse in the case above due to the "endgroup", which means that parts of executing `$SCRIPT` will incorrectly be hidden behind the "Display CPU and Memory information" group even though they were executed after the end of that group.

The runner should pipe both stdout and stderr into the same file descriptor, that is the only way to guarantee proper ordering (and it's what the shell does). Currently evidently it uses two different file descriptors here and that approach is fundamentally flawed.

**Comment by mwhipple** [comment]
Just came across this while looking for something else...but you should also be aware that by default stderr is not buffered whereas stdout is (with varying parameters (such as line buffering) depending on the shell mode which should at least be accounted for when doing local experiments in what is more likely an interactive session) - so independently of any GH specific implementation there is typically a need to change the buffering settings to keep stderr output from jumping ahead of some to stdout. The assertion that the shell uses the _same_ file descriptor could certainly use some substantiation but redirection is also a readily available option.

**Reply by RalfJung** [reply]
Stdout is typically buffered in a way that it gets flushed on newlines, so in practice interleaving of stdout and stderr lines works very well -- except on GH.

**Reply by mwhipple** [reply]
line buffered for interactive which GH actions would not be (as briefly mentioned above):
https://www.gnu.org/software/libc/manual/html_node/Buffering-Concepts.html (there are likely better references that I don't have on hand like: https://linux.die.net/man/3/stderr)
in any case this report seems to have a lot of assumptions that could use support.

**Reply by bjorn3** [reply]
Rust always line buffers stdout: https://github.com/rust-lang/rust/blob/6cf068db566de080dfa7ed24a216ea3aed2b98ce/library/std/src/io/stdio.rs#L589-L592

Also does Github Actions not use a tty for stdout/stderr? A lot of programs don't show any colors at all when stdout or stderr is not a tty.

**Reply by RalfJung** [reply]
>  in any case this report seems to have a lot of assumptions that could use support.

Well, for now we fixed our interleaving issues by redirecting stdout to stderr (or vice versa, I am not sure). So I think we have evidence that using the same FD for both helps, and stdout buffering is not an issue.

**Reply by mwhipple** [reply]
there's no argument that using the same FD helps which is where redirection was mentioned...
the assertion that that is standard behavior that GH deviates from is the premise for this issue which should be substantiated
there is no provided evidence that stdout buffering is the issue...there may be some confusion in the fact that the buffering is typically applied by the OS not within the application process (so its not a convention around how the code behaves but rather default behaviors of the two handles themselves).

For additional context I've worked through this behavior in a variety of tools and platforms since I default to using stdout and stderr and I've had to make adjustments where I cared about the ordering. I don't know specifically how GH operates and there may be some additional considerations, but I've run into this outside using a range of outputs and combinations of pipelines and individual applications written in different languages.

**Reply by RalfJung** [reply]
The script works fine when I run it on my laptop. It works fine on GHA with FD redirecting. It is broken on GHA in the default configuration.

IMO that is sufficient evidence to call GHA's configuration non-standard, and causing breakage. There may be other issues when the application does buffering the wrong way, but Rust programs seem to get that part right, making the underlying GHA issue visible.

**Reply by mwhipple** [reply]
Here's a good write-up:
http://www.pixelbeat.org/programming/stdio_buffering/

**Reply by mwhipple** [reply]
A basic illustrative example of the behavior:
```
#!/bin/bash

echo 'stdout'
echo 'stderr' >&2
```
execute the above interactively:
```
./test.sh
stdout
stderr
```
provide basic way to flip to more of a batch mode:
```
./test.sh | cat
stderr
stdout
```

**Reply by mwhipple** [reply]
To be comprehensive:
```
cat src/main.rs
fn main() {
    println!("stdout");
    eprintln!("stderr");
}
```
```
./target/debug/rustout
stdout
stderr
./target/debug/rustout | cat
stderr
stdout
```

None of the above is to suggest that there may not be other pieces in the mix that contribute, but the assertion that this behavior is a non-standard defect rather than something that should be expected and worked with is tenuous.

**Reply by RalfJung** [reply]
Yeah, with `| cat` stdout and stderr are not the same FD any more. But usually when you just run a program on the terminal, they are -- you can run `rustout` as often as you want, it will always print `stdout` and then `stderr`. And IMO GHA should be configured in a way to match the usual terminal behavior.

**Reply by mwhipple** [reply]
That's certainly a valid request but is different than suggesting a fundamental defect.
I personally would certainly _not_ want them to be the same since it's far easier to merge them using standard redirection (i.e. `exec 2>&1`) than it would be to try to tease them back apart but that's subjective.

**Reply by RalfJung** [reply]
> I personally would certainly not want them to be the same since it's far easier to merge them using standard redirection (i.e. exec 2>&1) than it would be to try to tease them back apart but that's subjective.

Github only provides the unified stdout/stderr log, right? So currently there's already no way to tease them back apart. It's the worst of both worlds.

**Reply by mwhipple** [reply]
the cat examples are potentially misleading since they don't redirect stderr :facepalm: ...I'm far more used to resolving this issue than trying to recreate it. A more straightforward reproduction would be to run the process in a container:
```
/rustout
stdout
stderr
docker run -v $(pwd):/host rust /host/rustout
stderr
stdout
```
as an additional note the latter invocation doesn't produce stable results:
```
for i in {1..10}; do docker run -v $(pwd):/host rust /host/rustout && echo '---'; done
stdout
stderr
---
stderr
stdout
---
stderr
stdout
---
stdout
stderr
---
stdout
stderr
---
stderr
stdout
---
stdout
stderr
---
stdout
stderr
---
stderr
stdout
---
stderr
stdout
---
```

The idea of redirecting the final sink (the log window) to a single descriptor could presumably help in this specific case, but wouldn't help if inversions are introduced earlier in a pipeline...so while the provided standard redirection mechanisms provide a means to merge streams when preserved ordering across them is desired; merging at the time out of output is fragile and implicitly forcing it earlier in the process is likely to deviate from standard UNIX-y behavior and could inhibit use of the separate streams.
If there's a desire to have a single ordered stream the mechanisms exist to realize that.

Without deeper explanation this behavior is in line with other more headless environments that I've used and so deviating from that by default may do more harm than good, though providing optional more attached style behavior could certainly be convenient for some cases.
